pkg/agent/openflow/network_policy.go

```diff
@@ -2220,3 +2228,64 @@ func (f *featureNetworkPolicy) l7NPTrafficControlFlows() []binding.Flow {
 			Done(),
 	}
 }
+
+func (f *featureNetworkPolicy) loggingNPPacketFlowWithOperations(cookieID uint64, operations uint8) binding.Flow {
+	loggingOperations := binding.NewRegMark(PacketInOperationField, uint32(operations))
+	fb := L2ForwardingOutTable.ofTable.BuildFlow(priorityNormal).
+		MatchRegMark(SendToControllerRegMark, loggingOperations)
+	if f.ovsMetersAreSupported {
+		fb = fb.Action().Meter(PacketInMeterIDNP)
+	}
+	return fb.Action().SendToController([]byte{uint8(PacketInCategoryNP), operations}, false).
+		Cookie(cookieID).
+		Done()
+}
+
+func (f *featureNetworkPolicy) initLoggingFlows() []binding.Flow {
+	flows := make([]binding.Flow, 0)
+	if !f.enableAntreaPolicy {
+		return flows
+	}
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	for operation := uint8(1); operation <= PacketInNPLoggingOperation+PacketInNPStoreDenyOperation+PacketInNPRejectOperation; operation++ {
+		flows = append(flows, f.loggingNPPacketFlowWithOperations(cookieID, operation))
+	}
+	return flows
+}
+
+func (f *featureNetworkPolicy) loggingGroups() []binding.OFEntry {
+	var groups []binding.OFEntry
+	if !f.enableAntreaPolicy {
+		return groups
+	}
+	candidateTables := []*Table{EgressRuleTable, EgressMetricTable, IngressRuleTable, IngressMetricTable}
+	if f.enableMulticast {
+		candidateTables = append(candidateTables, MulticastEgressMetricTable, MulticastIngressMetricTable)
+	}
+	for _, nextTable := range candidateTables {
+		groupKey := fmt.Sprintf("%s%d", loggingAndResubmitTableGroupPrefix, nextTable.GetID())
+		obj, ok := f.loggingGroupCache.Load(groupKey)
+		if ok {
+			groups = append(groups, obj.(binding.Group))
+			continue
+		}
+		// Create OpenFlow group to both log ANP and resubmit the packet to nextTable.
+		groupID := f.groupAllocator.Allocate()
+		group := f.bridge.NewGroupTypeAll(groupID).Bucket().
```

Comment:
So I assume this is where the packet fork happens? Could we add a bit more comment here on what each bucket does?

Reply:
You are correct, added comments.